vizro-core/src/vizro/models/_action/_action.py

```diff
@@ -106,32 +94,42 @@ def _get_callback_mapping(self):
 
         return callback_inputs, callback_outputs, action_components
 
-    def _action_callback_function(self, **inputs: Dict[str, Any]) -> Dict[str, Any]:
+    def _action_callback_function(self, inputs: Dict[str, Any], callback_outputs: Dict[str, Any]) -> Dict[str, Any]:
         logger.debug("=============== ACTION ===============")
         logger.debug(f'Action ID: "{self.id}"')
         logger.debug(f'Action name: "{self.function._function.__name__}"')
         logger.debug(f"Action inputs: {inputs}")
 
         # Invoking the action's function
-        return_value = self.function(**inputs) or {}
+        return_value = self.function(**inputs)
 
         # Action callback outputs
-        outputs = list(ctx.outputs_grouping.keys())
+        outputs = list(callback_outputs.keys())
         outputs.remove("action_finished")
```

Comment:
We need it since the `callback_outputs` contains all the action callback outputs:
```
        callback_outputs: Dict[str, Any] = {
            **_get_action_callback_mapping(action_id=ModelID(str(self.id)), argument="outputs"),
            **{
                f'{output.split(".")[0]}_{output.split(".")[1]}': Output(
                    output.split(".")[0], output.split(".")[1], allow_duplicate=True
                )
                for output in self.outputs
            },
            "action_finished": Output("action_finished", "data", allow_duplicate=True),
        }
```

We can avoid it by adding "action_finished" key inside the `build` method except inside the `self._get_callback_mapping`, but I don't think we can gain any value from this move.

Reply:
Ah ok. Not a big deal but I think this would be cleaner actually if you don't mind making the change:

```
 @callback(output=callback_outputs, inputs=callback_inputs, prevent_initial_call=True)
        def callback_wrapper(trigger: None, **inputs: Dict[str, Any]) -> Dict[str, Any]:
            outputs = list(callback_outputs)
            outputs.remove("action_finished")
            return_dict =  self._action_callback_function(inputs=inputs, outputs=outputs)
            return {"action_finished": None, **return_dict}
```

Then `_action_callback_function` can become even closer to "pure" execution of the callback function that just takes in `outputs: List[str]`.